[Frontend][Publisher] Put agency ID in the url, add search params for metrics (Data page)

ivan-kishko opened the pull request:
## Description of the change

> Add agency ID in the URL (as url param)
Add search params for metrics in data page
Fix some bugs (data view in data page)
Change some tests

## Type: Feature

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

Closes #149 #150 #170

## Checklists

### Development

**This box MUST be checked by the submitter prior to merging**:
- [x] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [x] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

ivan-kishko commented:
Hello to all! Just want to comment something out before you start a review.
1. With this PR I did #149 and #150 because it would slow down things a lot for #150 without #149
2. This is a big PR. But anyway it's been a while, sorry for that. And I really tried to find all possible edge cases since user can now manually enter agencyId in url, but probably you will find some bugs, sorry for that too. Will try to fix all of them as fast as possible.
3. I did some fixes related to my previous task (data view), there was a bug when you load page on small screen chart was not displayed and a bug with insights not disappear in some cases as well or disappear incorrectly.
4. You will notice that I commented out one test completely, because now <code>userInfoLoaded</code> flag is used once inside <code>App</code> component.
I guess that is all I wanted to say for now. Waiting for your comments! Thanks.

lilidworkin commented:
Super exciting @ivan-kishko ! Looking forward to reviewing.

ivan-kishko commented:
Hello all again) Forgot a couple of things! I guess I need some design on how to make page which says that "user has no agency with given ID" and probably I will add a commit for #170 in this PR as well as it is url problem.

lilidworkin commented on publisher/src/App.tsx:
```diff
@@ -17,48 +17,56 @@
+      </PageWrapper>
+    );
+
+  // using this variable to indicate whether user has any agencies
```
Thanks for this helpful comment!

lilidworkin commented on publisher/src/App.tsx:
```diff
@@ -17,48 +17,56 @@
-            path={`/${REPORTS_LOWERCASE}/create`}
-            element={<CreateReports />}
+            path="/"
+            element={<Navigate to={`/agency/${initialAgency}/reports`} />}
```
We're trying to get rid of the word "reports" -- let's change this to `/agency/${initialAgency}/${REPORTS_LOWERCASE}` (cc @mxosman to check me on that)

ivan-kishko replied:
done [6fc8886](https://github.com/Recidiviz/justice-counts/pull/187/commits/6fc888678bc14fae95118615b5cc76fbad78cd08)

lilidworkin commented on publisher/src/App.tsx:
```diff
@@ -17,48 +17,56 @@
+            element={<Navigate to={`/agency/${initialAgency}/reports`} />}
           />
+          <Route path="/agency/:agencyId/*" element={<Router />} />
+          {/* TBD how to treat random routes */}
```
I think the behavior you have here -- routing to /reports page -- is great!

lilidworkin replied:
Or actually, once we have the designs 404 page (they're coming) we could go there instead.

lilidworkin replied:
We can remove this TBD comment -- I think the behavior you have is good!

lilidworkin commented on publisher/src/App.tsx:
```diff
@@ -17,48 +17,56 @@
-            path={`/${REPORTS_LOWERCASE}/:id`}
-            element={<ReportDataEntry />}
+            path="*"
+            element={<Navigate to={`/agency/${initialAgency}/reports`} />}
```
As above, let's make this `/agency/${initialAgency}/${REPORTS_LOWERCASE}`

ivan-kishko replied:
done [6fc8886](https://github.com/Recidiviz/justice-counts/pull/187/commits/6fc888678bc14fae95118615b5cc76fbad78cd08)

lilidworkin commented on publisher/src/components/Menu/Menu.tsx:
```diff
@@ -46,9 +46,12 @@ const Menu = () => {
   const navigate = useNavigate();
   const location = useLocation();
 
+  const pathWithoutAgency = location.pathname.split("/").slice(3).join("/");
```
I wonder if there's a better way to do this? This feels a little fragile/brittle to me. But not a big deal -- it works! @mxosman , @terryttsai  lmk if you guys have any thoughts.

mxosman replied:
Love the helper!

lilidworkin commented on publisher/src/components/MetricConfiguration/MetricConfiguration.tsx:
```diff
@@ -64,51 +65,47 @@ export const MetricConfiguration: React.FC = observer(() => {
-        : userStore.currentAgency?.systems[0];
       const isUrlMetricParamInCurrentSystem = !!getMetricsBySystem(
-        systemToCheckMetrics
+        systemSearchParam
```
Why was this changed? I think maybe this part should stay the same?

ivan-kishko replied:
Basically, code get to this point ONLY if systemSearchParam from external link belongs to agency in the url. Thats why there is no point to check if systemSearchParam from external link belongs to agency again and therefore no reason to check metric key in first system of the agency since in case if systemSearchParam DOES NOT belongs to an agency user will be redirected to metric config page with first available system.

lilidworkin replied:
Oh I see, you're right! Thanks for explaining.

lilidworkin commented on publisher/src/components/MetricConfiguration/MetricConfiguration.tsx:
```diff
@@ -64,51 +65,47 @@ export const MetricConfiguration: React.FC = observer(() => {
       )?.find((metric) => metric.key === metricSearchParam);
       if (!isUrlMetricParamInCurrentSystem) {
-        setSettingsSearchParams({});
+        setSettingsSearchParams({ system: currentAgency?.systems[0] });
```
Same here -- not obvious to me why this was changed. I might just be missing something!

ivan-kishko replied:
actually maybe need to just make metric param undefined - Ill fix this

ivan-kishko replied:
or as in MetricsView show message and propose to reload page with default params

lilidworkin commented on publisher/src/components/MetricConfiguration/MetricsView.tsx:
```diff
@@ -49,166 +47,157 @@ import {
+    // same logic as in metric config page, the only difference is
+    // there should always be metric search param
+    // -------------------------------------------
+    // Maybe instead of checking system and metric belongings better to show
```
Yea I actually really like that idea! Because this logic is pretty confusing (not your fault, it was before too). And generally I like the idea of showing explicit messages instead of trying to be clever with our redirects. Curious what @mxosman and @terryttsai  think.

ivan-kishko replied:
I guess then really need some design)

lilidworkin replied:
Yea totally. You can hold off on this for now until we hear back from some other folks!

mxosman replied:
I'm not sure I 100% follow -- would this be when someone enters a bad `system` or `metric` in the URL? I personally don't mind the behavior of redirecting to the default system & metric vs. showing a page that asks them to reload. I can totally see the case made for either/or.

mxosman replied:
Chatted with Ivan briefly yesterday and I'm convinced it's better to not let the user assume their bad entry was correct and have some kind of explicit messaging like you both suggested.

lilidworkin commented on publisher/src/components/Reports/ReportDataEntry.tsx:
```diff
@@ -135,25 +135,22 @@ const ReportDataEntry = () => {
+  useEffect(() => {
+    // in case user type (change) agency inside url by hand
+    if (!reportStore.reportOverviews[reportID]) {
+      navigate(`/agency/${agencyId}/reports`);
```
I think it might be better to show a blank page with a "Report X Not Found" message in this case? I think in general, I'd rather show explicit messages than try to redirect. That might cause a confusing experience and also make it hard to debug things.

lilidworkin commented on publisher/src/components/Settings/utils.ts:
```diff
@@ -23,3 +23,6 @@ export const getActiveSystemMetricKey = ({
   return `${system?.toUpperCase()}-${metric}`;
 };
+
+export const getSettingsMenuItemFromLocation = (location: string) =>
```
Also curious if there's a better way to do this rather than string parsing, but fine if not!

lilidworkin commented on publisher/src/pages/InvalidAgency.tsx:
```diff
@@ -0,0 +1,62 @@
+  cursor: pointer;
+`;
+
+// TODO need figma design for this page
```
I wonder if we should just design a generic 404 Not Found page, and then use that in all of these various places? I'll ask Humphrey/Juan.

lilidworkin commented on publisher/src/stores/UserStore.ts:
```diff
@@ -152,6 +131,15 @@ class UserStore {
 
+  getCurrentAgency(agencyId: string | undefined): UserAgency | undefined {
+    if (agencyId) {
+      return this.userAgencies?.find(
```
This is a nit, but since we call this function a lot, rather than iterating through the list each time (which is O(n)), can we use a map or dictionary data structure for O(1) lookup?

ivan-kishko replied:
added userAgenciesById to userStore [6fc8886](https://github.com/Recidiviz/justice-counts/pull/187/commits/6fc888678bc14fae95118615b5cc76fbad78cd08)

lilidworkin commented on publisher/src/components/Menu/Menu.tsx:
```diff
@@ -134,9 +140,10 @@ const Menu = () => {
                     key={agency.id}
                     onClick={() => {
-                      userStore.setCurrentAgencyId(agency.id);
+                      // userStore.setCurrentAgencyId(agency.id);
```
We can just delete this line! I agree that we don't need `setCurrentAgencyId` anymore.

lilidworkin reviewed (commented):
Wow @ivan-kishko , this is amazing! I knew this would be a lot of work but I didn't quite realize HOW MUCH. Incredible job!

I left some comments, but also want to test this out more later. Overall it looks really great.

> Probably might remove current agency check in store methods but want to hear your opinion on it

Can you explain this a bit more? Which check are you talking about?

ivan-kishko commented:
> Can you explain this a bit more? Which check are you talking about?

Hello @lilidworkin. So by this I meant if you look at for example ReportStore... We have here getReportOverviews, createReport, initializeReportSettings, updateReportSettings, getUploadedFilesList - at the beginning there is a check if currentAgency is tied to a user. With current routing we are checking if agency with given id (in the url) is tied on site load and if not we display a message to contact support. Maybe I am missing something but If we check it on site load maybe we do not need this check inside this requests?

terryttsai commented:
Since we're both developing and modifying DatapointsView.tsx to suit our different apps, there's again merge conflicts—don't worry about the merge conflicts. I think I'll resolve them again in your PR, and if this keeps happening I'm slowly breaking out the metric title and metric insights and hopefully that'll help prevent these conflicts in the future.

lilidworkin commented on publisher/src/components/Reports/DataEntryForm.test.tsx:
```diff
@@ -47,23 +49,25 @@ test("display loading when no reports are loaded", async () => {
-  expect(errorText).toBeInTheDocument();
-  expect.hasAssertions();
-});
+// test("display error when report fails to load", async () => {
```
Why is this test failing?

ivan-kishko replied:
I am not sure for now, currently investigating, all other test failed due to wrong url inside tests itself

mxosman replied:
Let me know if you need help - I imagine some of these old tests will be a bit tricky especially since so much has changed.

ivan-kishko replied:
fixed with [71a58a3](https://github.com/Recidiviz/justice-counts/pull/187/commits/71a58a330bde1c24433caa03a32c5d12b88087d6) - actually it was not test that has to be fixed but ReportDataEntry component

lilidworkin commented on publisher/src/router/Router.tsx:
```diff
@@ -0,0 +1,79 @@
+          <Route path="/settings/*" element={<Settings />} />
+          <Route path="/upload" element={<DataUpload />} />
+          <Route path="/upload/review-metrics" element={<ReviewMetrics />} />
+          {/* TBD how to treat random routes */}
```
Once we have a 404 page, we could go there instead!

lilidworkin replied:
Ok so I think this is actually fine for now -- we can remove the TBD comment.

lilidworkin commented on publisher/src/utils/helperUtils.ts:
```diff
@@ -222,3 +222,13 @@ export function memoizeDebounce<F extends (...args: any[]) => any>(
 }
+
+/**
+ * remove agency part (/agency/agencyId/) from location pathname
```
love this, thank you!

lilidworkin reviewed (commented):
> Maybe I am missing something but If we check it on site load maybe we do not need this check inside this requests?

You're definitely right! You can remove the `const currentAgency = this.userStore.getCurrentAgency(agencyId);` from all those store methods. Since you're now passing in `agencyId` as a parameter, that should be all you need. Good catch!

lilidworkin commented:
@ivan-kishko while testing this locally, I was able to get into a weird state where the Metric Configuration page is blank. But I think it's possible that @mxosman already fixed this in another PR? I'm not sure. Here's a video:

https://user-images.githubusercontent.com/99899591/205108901-bfc2001a-91c6-46ae-b591-2ab9dcec92cf.mov

lilidworkin commented:
One other small thing I found while testing: On the Data Entry page, the link in "Additional context can be viewed and entered via the Metric Configuration page in Settings." seems to take you to the Reports page, not Metric Configuration.

ivan-kishko commented:
> One other small thing I found while testing: On the Data Entry page, the link in "Additional context can be viewed and entered via the Metric Configuration page in Settings." seems to take you to the Reports page, not Metric Configuration.

yeah this is just 1 of many links that i missed to fix. Will be fixed in next commit

mxosman commented:
> @ivan-kishko while testing this locally, I was able to get into a weird state where the Metric Configuration page is blank. But I think it's possible that @mxosman already fixed this in another PR? I'm not sure. Here's a video:
> 

You're right, Lili! It was fixed in #182.

Looking forward to reviewing all of this incredible work today, Ivan!

mxosman commented on publisher/src/components/MetricConfiguration/MetricConfiguration.tsx:
```diff
@@ -64,51 +65,47 @@ export const MetricConfiguration: React.FC = observer(() => {
+        system: currentAgency?.systems[0],
       });
     }
+    setIsLoading(false);
```
Thank you for all the comments!

mxosman commented on publisher/src/components/MetricConfiguration/MetricConfiguration.tsx:
```diff
@@ -117,35 +114,15 @@ export const MetricConfiguration: React.FC = observer(() => {
     // eslint-disable-next-line react-hooks/exhaustive-deps
-    [userStore]
-  );
+  }, [agencyId]);
```
This cleaned so much up!

mxosman commented on publisher/src/components/MetricConfiguration/RaceEthnicitiesForm.tsx:
```diff
@@ -239,7 +242,7 @@ export const RaceEthnicitiesForm = observer(() => {
                         raceEthnicityGridStates
                       );
-                      saveMetricSettings(updatedDimensions);
+                      saveMetricSettings(updatedDimensions, agencyId);
```
I'm sure we'll have a conflict in this file - these all now use a debounced save. Let me know if resolving the conflict isn't straightforward!

mxosman commented on publisher/src/components/MetricConfiguration/MetricsView.tsx:
```diff
@@ -221,13 +210,7 @@ export const MetricsView: React.FC = observer(() => {
-                      pathname: "/settings/metric-config",
-                      search: `?${createSearchParams(params)}`,
-                    });
+                    navigate("../settings/metric-config");
```
Will this need to be updated?

ivan-kishko replied:
no. It is relative pathing in most places now since I did nested routing. probably for readability If You want I can change all this paths (except in Router) to absolute ones

mxosman replied:
No need - thank you for clarifying this!

mxosman commented on publisher/src/components/Reports/ReportDataEntry.tsx:
```diff
@@ -98,62 +82,20 @@ const ReportDataEntry = () => {
     // eslint-disable-next-line react-hooks/exhaustive-deps
-    []
-  );
+  }, []);
```
Beautifully simplified! Thank you for dealing with my past-self's mess!

mxosman commented on publisher/src/components/Reports/ReportDataEntry.tsx:
```diff
@@ -178,56 +120,48 @@ const ReportDataEntry = () => {
+        reportID={reportID}
+        activeMetric={activeMetric}
+        fieldDescription={fieldDescription}
+      />
```
Love that the confirmation is in its own route now!

mxosman reviewed (commented):
This was a beast of an undertaking - it looks and feels fantastic!

One other small thing I noticed was the styling that indicates the active setting link you're in is missing:

<img width="387" alt="Screenshot 2022-12-01 at 11 25 54 AM" src="https://user-images.githubusercontent.com/59492998/205156233-b236728d-a389-4376-bda1-d030bb06a2ef.png">

Should be:
<img width="299" alt="Screenshot 2022-12-01 at 2 48 28 PM" src="https://user-images.githubusercontent.com/59492998/205156372-1845a82e-2946-451b-a222-c342af20d147.png">

ivan-kishko commented:
> This was a beast of an undertaking - it looks and feels fantastic!
> 
> One other small thing I noticed was the styling that indicates the active setting link you're in is missing:
> 
> <img alt="Screenshot 2022-12-01 at 11 25 54 AM" width="387" src="https://user-images.githubusercontent.com/59492998/205156233-b236728d-a389-4376-bda1-d030bb06a2ef.png">
> 
> Should be: <img alt="Screenshot 2022-12-01 at 2 48 28 PM" width="299" src="https://user-images.githubusercontent.com/59492998/205156372-1845a82e-2946-451b-a222-c342af20d147.png">

Hello @mxosman. Yes I noticed that too - will be fixed in next commit, I already made changes)

ivan-kishko commented on common/components/DataViz/DatapointsTitle.tsx:
```diff
@@ -36,11 +36,14 @@ export const DatapointsTitle: React.FC<{
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [metricName]
+  );
```
Thanks! Well actually I noticed that it was buggy, like if you open data page initially on small page it would behave completely wrong, so I started googling and found that solution. The tricky part was that if display:none there is nothing to measure so I needed to change css a bit as well for it to start working as intended.

ivan-kishko commented on publisher/src/pages/Reports.tsx:
```diff
@@ -337,7 +311,11 @@ const Reports: React.FC = () => {
+                            reportStore.deleteReports(
+                              reportsToDelete,
+                              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                              agencyId!
```
before all this refactor this functions were taking current agency id from userStore. Since now it is removed, and actually you cant enter any page if agencyId is invalid or user does not have agencies at all it looks to me like a safe option. I think it is the only or 1 of 2-3 functions that has agencyId! like that. I dont really like that either but if you check on this page current agency it would just a be a double check since first is dine on site load

lilidworkin replied:
I think this is totally fine. Thanks for explaining Ivan!

mxosman replied:
Thanks for explaining, Ivan!

ivan-kishko commented on publisher/src/components/Header/Header.tsx:
```diff
@@ -16,18 +16,30 @@
+      <LogoContainer
+        onClick={() =>
+          navigate(
+            isAgencyValid ? `/agency/${agencyId}/${REPORTS_LOWERCASE}` : "/"
```
Hello @terryttsai. It will go to /agency/userAgencies[0]/records

ivan-kishko replied:
actually I can make it more readable by taking initial agency and write a path like /agency/${initialAgencyId}/records

ivan-kishko commented on publisher/src/components/Menu/Menu.tsx:
```diff
@@ -72,38 +76,41 @@ const Menu = () => {
       setActiveMenuItem(MenuItems.Reports);
-    } else if (location.pathname === `/${REPORTS_LOWERCASE}/create`) {
+    } else if (location.pathname === `${REPORTS_LOWERCASE}/create`) {
       setActiveMenuItem(MenuItems.CreateReport);
```
it is not. I removed it but then I saw it while was doing first main merge into my brach so I thought maybe you left it there for a reason

ivan-kishko commented on publisher/src/components/Menu/Menu.tsx:
```diff
@@ -72,38 +76,41 @@ const Menu = () => {
       setActiveMenuItem(MenuItems.CreateReport);
-    } else if (location.pathname === "/settings") {
+    } else if (pathWithoutAgency === "settings") {
       setActiveMenuItem(MenuItems.Settings);
```
Yeah, that is because we do not have such pathname, it is either settings/account(or /uploads or /metric-config), will fix it!

ivan-kishko commented on publisher/src/components/Menu/Menu.tsx:
```diff
@@ -72,38 +76,41 @@ const Menu = () => {
       setActiveMenuItem(MenuItems.Settings);
-    } else if (location.pathname === "/data") {
+    } else if (pathWithoutAgency === "data") {
       setActiveMenuItem(MenuItems.Data);
```
Well actually now there can only be 3 active items. I can completely remove useEffect useState, active menu item and just write inside MenuItem something like (removeAgencyFromPath === "data"/"records"/"settings"). 3 items is not that much, what do you think?

ivan-kishko replied:
[976863f](https://github.com/Recidiviz/justice-counts/pull/187/commits/976863f9b6e2b3bd5fefb3b00e688f001e2ed64c) done here

terryttsai replied:
Looks awesome!

ivan-kishko commented:
General question. Now with React 18 devs are working on a feature that preserve state after component unmounts. while developing actual functionality for now useEffect with empty array called twice (as they say it is a stress test). It is though as far as I read, only in strict mode and dev env. This should not be a problem on production. You can feel it if you go on data page and then try to go right after in settings for example - it will switch back to data page anyway. Should I do something with it or leave it like that?

terryttsai commented on publisher/src/stores/MetricConfigStore.tsx:
```diff
@@ -197,9 +199,10 @@ class MetricConfigStore {
+    agencyId: string | undefined
   ): Promise<Response> => {
-    const { currentAgency } = this.userStore;
+    const currentAgency = this.userStore.getCurrentAgency(agencyId);
```
We could rename this method `getCurrentAgency` to `getAgency` since it's just getting any agency by agencyId

ivan-kishko replied:
[71b3b30](https://github.com/Recidiviz/justice-counts/pull/187/commits/71b3b30c434bf6f3cbd6dfe625774e29eff8f112) done!

ivan-kishko commented:
Hello to all once more @lilidworkin @terryttsai @mxosman )! So reading all this comments from you guys looks like there is no agreement on how to treat wrong searchParams. I guess You need to decide what to do - and we need to use this decision in all places. Either if something wrong just show generic error page with proposition to RELOAD page with valid params or LINK TO homepage - OR automatically replace invalid search params with valid ones. (If first I have to wait for some kind of design) (AND I guess I am repeating myself asking about this but still).
Second - all this methods from stores that are checking if agency undefined or not. Since this check is done on site load and  You @lilidworkin said that I can get rid of these, I want to make sure that all of You feel OK that in a lot of places agency id will be passed like this "agencyId!" with <code>// eslint-disable-next-line @typescript-eslint/no-non-null-assertion</code>.
Looking forward to hear from You!)

lilidworkin commented on publisher/src/pages/NoAgencies.tsx:
```diff
@@ -0,0 +1,110 @@
+        </LogoBlock>
+        <TitleBlock>No agency has been connected to this account.</TitleBlock>
+        <InfoBlock>
+          Please reach out to{" "}
```
Let's change this to "Please email the Justice Counts team at  <a href="mailto:justice-counts-support@csg.org">justice-counts-support@csg.org</a> for support."

lilidworkin commented on publisher/src/pages/NotFound.tsx:
```diff
@@ -0,0 +1,66 @@
+      <Title>{title}</Title>{" "}
+      <HomepageLinkWrapper>
+        Go to{" "}
+        <HomepageLink onClick={() => navigate(pathname)}>homepage</HomepageLink>
```
Let's change this to "Go back <HomepageLink onClick={() => navigate(pathname)}>home</HomepageLink>."

lilidworkin commented on publisher/src/pages/NotFound.tsx:
```diff
@@ -0,0 +1,66 @@
+  cursor: pointer;
+`;
+
+// TODO need figma design for this page
```
We're working on it! I don't think you need to wait for it. The page looks fine for now and we can make it prettier in a follow-up PR!

lilidworkin commented on publisher/src/router/Router.tsx:
```diff
@@ -0,0 +1,79 @@
+            path={`/${REPORTS_LOWERCASE}/:id/review`}
+            element={<ReviewReportDataEntry />}
+          />
+          <Route path="/settings/*" element={<Settings />} />
```
If I go to `http://localhost:3000/agency/154/settings/xxx`, I just get a blank page. Not a big deal, but maybe there's a better way to handle?
<img width="1113" alt="Screen Shot 2022-12-02 at 11 26 24 AM" src="https://user-images.githubusercontent.com/99899591/205339173-952b201e-535b-4efb-8de0-bcad376688e0.png">

lilidworkin replied:
Yes that would be great! Thanks

lilidworkin reviewed (commented):
Thanks for all of your hard work and patience on this, Ivan! A few things:

**1. Regarding how to treat wrong search params:** Yea, sorry for the mixed messages here! I think your approach of _redirecting the user to valid params is fine_. I do think we should be _consistent_ with how we do this throughout the app, and do it the _same_ way on the Data page and the Settings page (if that's possible). I'm not sure I'm fully understanding @terryttsai 's comment [here](https://github.com/Recidiviz/justice-counts/pull/187#discussion_r1037667362), but maybe @ivan-kishko and @terryttsai can discuss?

**2. Regarding using agencyId!:** I am 99% sure this is fine, because like Ivan says, it's impossible for you to be in the app with a null agencyId. Unless @mxosman or @terryttsai very much feel differently, I think we can remove the checks for null agencyId and just use force it with !.

**3. Regarding Ivan's comment [here](https://github.com/Recidiviz/justice-counts/pull/187#issuecomment-1334547023)**: I think if this _does not affect production_ then this is fine and I wouldn't worry about it. Curious if @mxosman or @terryttsai have other thoughts.

**4. Spinner re-running?** I noticed that when the pages load, the spinner will start spinning, then stop and then start again. I'm curious why this happens? It's not a big deal, but I wonder if there's some extra/redundant work happening behind the scenes? See video:


https://user-images.githubusercontent.com/99899591/205340940-62d97cb8-c5d5-498d-bdff-e5d2228d8b8c.mov

ivan-kishko commented:
> Thanks for all of your hard work and patience on this, Ivan! A few things:
> 
> **1. Regarding how to treat wrong search params:** Yea, sorry for the mixed messages here! I think your approach of _redirecting the user to valid params is fine_. I do think we should be _consistent_ with how we do this throughout the app, and do it the _same_ way on the Data page and the Settings page (if that's possible). I'm not sure I'm fully understanding @terryttsai 's comment [here](https://github.com/Recidiviz/justice-counts/pull/187#discussion_r1037667362), but maybe @ivan-kishko and @terryttsai can discuss?
> 
> **2. Regarding using agencyId!:** I am 99% sure this is fine, because like Ivan says, it's impossible for you to be in the app with a null agencyId. Unless @mxosman or @terryttsai very much feel differently, I think we can remove the checks for null agencyId and just use force it with !.
> 
> **3. Regarding Ivan's comment [here](https://github.com/Recidiviz/justice-counts/pull/187#issuecomment-1334547023)**: I think if this _does not affect production_ then this is fine and I wouldn't worry about it. Curious if @mxosman or @terryttsai have other thoughts.
> 
> **4. Spinner re-running?** I noticed that when the pages load, the spinner will start spinning, then stop and then start again. I'm curious why this happens? It's not a big deal, but I wonder if there's some extra/redundant work happening behind the scenes? See video:
> 
>  Screen.Recording.2022-12-02.at.11.35.05.AM.mov 
> **5. Merge conflicts** @ivan-kishko do you want to take a stab at resolving the merge conflicts, and let us know what challenges you run into? Or would you prefer @mxosman or @terryttsai do it?

Hello @lilidworkin! Thanks for answers!) 
1. About conflicts I will try to do it by myself when I finish. 
2. About loader - If you enter site from scratch - 1st loader spins while user info loading (in App.tsx) and then second loader spins while data is loading for corresponding page. Looks like after initial load even though useEffect runs twice loader looks fine.

mxosman commented:
> **2. Regarding using agencyId!:** I am 99% sure this is fine, because like Ivan says, it's impossible for you to be in the app with a null agencyId. Unless @mxosman or @terryttsai very much feel differently, I think we can remove the checks for null agencyId and just use force it with !.
> 

I'm good to go with it - just a pattern I haven't seen before. Should be fine since the null agencyId case is being handled

> **3. Regarding Ivan's comment [here](https://github.com/Recidiviz/justice-counts/pull/187#issuecomment-1334547023)**: I think if this _does not affect production_ then this is fine and I wouldn't worry about it. Curious if @mxosman or @terryttsai have other thoughts.
> 

Yeah, I'm still fresh on the React 18 update... they did mention this as a heads up, but I haven't ran into weird issues with it in development, yet. Ivan - I can't reproduce going to the data entry page & switching to settings - I don't see the switch back to the data entry page even when I do this via URL. If it starts to become an issue we can revert back to React 17 that we had before with minimal-no consequences.

terryttsai commented:
> Regarding how to treat wrong search params: Yea, sorry for the mixed messages here! I think your approach of redirecting the user to valid params is fine. I do think we should be consistent with how we do this throughout the app, and do it the same way on the Data page and the Settings page (if that's possible). I'm not sure I'm fully understanding @terryttsai 's comment https://github.com/Recidiviz/justice-counts/pull/187#discussion_r1037667362, but maybe @ivan-kishko and @terryttsai can discuss?

Sorry for the confusion, I like the approach of redirecting the user to valid params and I'm cool with that approach! I was worried about that specific logic breaking in the future but typescript would probably catch it. I was also proposing a 3rd approach and that's just redirect the user to the homepage but that's probably a worse approach than the other two.

ivan-kishko commented:
Hello again! @mxosman @terryttsai can you please check out last 2 commits I did. Added message for user when redirecting to default search params, removed agencyId checks and fixed some tests. Thank you!

terryttsai commented:
The new code changes look good to me, @ivan-kishko would you like to try to resolve the merge conflicts or shall I go ahead and resolve them?

ivan-kishko commented:
Hello to all once again! So I guess this 2 commits will be final ones. @terryttsai please check if I get you right with metric insights. @mxosman please check those debounceSaves if I merged them correctly. And @lilidworkin wanted to take a final look, I think it is the time. Thank you guys in advance!)

ivan-kishko commented:
> @ivan-kishko the latest changes look great! I like the toast! Here's a video if @terryttsai or @mxosman want to take a look:
> 
>  Screen.Recording.2022-12-06.at.11.23.03.AM.mov 
> (I did notice that the toast seems to flash twice, but that's not a big deal.)
> 
> @ivan-kishko I think this is ready to go -- I wouldn't change any more functionality in this PR since it's already quite big. Can you try resolving the merge conflicts and let us know if you have any trouble? Once the conflicts are resolved, I'd like to test it again one more time before accepting. Thank you!

sadly flashes because of new useEffect I wrote about it above...

terryttsai commented:
Thanks for going through and resolving the merge conflicts! Thanks also for preserving your fix to the metric insights cut-off by hiding the metric insights at a certain width, I think that's probably the best solution for now until we get new designs for this.

lilidworkin commented:
Looks awesome! Let's merge this beast @ivan-kishko ! It's ready to go. If we need to fix anything else later, we can always do in a followup.

lilidworkin commented:
And thanks @terryttsai for the careful PR review!
